Fix WooPay user registration in classic checkout

malithsen opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/9552

#### Changes proposed in this Pull Request
This re-introduces a check that was removed in https://github.com/Automattic/woocommerce-payments/pull/9464 to fix WooPay user registration when registering through the "Save my info" checkbox in the classic checkout page.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Test classic checkout**
1. Add a product to the cart.
2. Go to classic checkout page.
3. Enter a WooPay unregistered email address.
4. Check the "Securely save my information for 1-click checkout" checkbox.
5. Place the order.
6. Go to WooPay admin area.
7. Visit WooPay login page (`woopay.test/account/login`)
8. You should be able to login with the email used in step 2.

**(Regression) Test Blocks checkout**
Repeat the above tests, but in step 2, use blocks checkout.
 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

malithsen committed:
Check $_POST attribute when creating WooPay user

- $_POST check is needed for shortcode checkout.
- exising session_data check is needed for blocks checkout

malithsen committed:
Add unit test

malithsen committed:
Add changelog entry

malithsen committed:
Merge branch 'develop' into fix/woopay-user-creation-in-shortcode-checkout

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9558` or branch name `fix/woopay-user-creation-in-shortcode-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/woopay-user-creation-in-shortcode-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: bed34443fdc2550d81e6f44d46fa94ffa589058b
- Build time: 2024-10-18 09:53:36 UTC

_Note: the build is updated when a new commit is pushed to this PR._

malithsen committed:
Limit test to classic checkout

a-danae commented on includes/woopay/class-woopay-utilities.php:
```diff
@@ -125,7 +125,7 @@ public function should_save_platform_customer() {
 		}
 
-		return isset( $session_data['save_user_in_woopay'] ) && filter_var( $session_data['save_user_in_woopay'], FILTER_VALIDATE_BOOLEAN );
+		return ( isset( $_POST['save_user_in_woopay'] ) && filter_var( wp_unslash( $_POST['save_user_in_woopay'] ), FILTER_VALIDATE_BOOLEAN ) ) || ( isset( $session_data['save_user_in_woopay'] ) && filter_var( $session_data['save_user_in_woopay'], FILTER_VALIDATE_BOOLEAN ) ); // phpcs:ignore WordPress.Security.NonceVerification
```
🤏 How about splitting this into variables so it's easier to read? It took me a bit to understand how the conditionals were grouped. Thinking that it'd be simpler if we could do `return $is_save_user_in_post || $is_save_user_in_session`. Disregard the variable names, just referring to readability.

Not a blocker, though!

a-danae reviewed (commented):
Thanks for fixing this, @malithsen ! Things look good to me! I do have a question before approving.

I was able to replicate the problem the issue in `develop`, the WooPay user isn't being created when checking off the "Save my information" checkbox and placing the order through the shortcode checkout. I could see that the method `should_save_platform_customer()` was returning `false`

Then, in this branch:
✅ `should_save_platform_customer()` does return `true`.
✅  The WooPay user does get created when placing the order through the block checkout when checking off the "Save my info" checkbox.
✅  The code changes make sense to me. I left a comment, but it's not a blocker.

❓ I wasn't able to get the WooPay user created when placing the order through the shortcode checkout. This is what I'm seeing:

- The WCPay server has the following entry in the error log 
   ```
   PHP Warning:  Undefined array key "platform_checkout_phone" in /var/www/html/wp-content/plugins/woocommerce-payments-server/server/wp-content/rest-api-plugins/endpoints/wcpay/woopay/service/class-woopay-service.php on line 258
  ```

- This `platform_checkout_phone` seems to be set in the metadata on WooPayments, [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/woopay-user/class-woopay-save-user.php#L88). This value is empty when using the shortcode checkout.
- The phone is retrieved from the Session data, [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/woopay/class-woopay-utilities.php#L165-L178), which doesn't seem to be set for the shortcode checkout.

The phone is set under the same array we're checking in the session but within `$_POST`, so the fix could be retrieving it from POST when not set in Session. But I'm curious why user creation is working for you and not for me. 

Either way, it may be worth checking for POST too, in case this empty session behavior happens in real stores. What do you think?

malithsen committed:
Check for $_POST vars when creating WooPay user.

When WooPay user is created via shortcode checkout, some of the metadata
are coming through $_POST. This makes sure we capture those.

malithsen commented:
@a-danae Thanks for the review Danae and for catching the missing phone number!

> But I'm curious why user creation is working for you and not for me.

My guess is that, I had the phone number already set on my session, possibly when I navigated to the blocks checkout page. 
In any case, I was able to recreate the issue you described. 

In https://github.com/Automattic/woocommerce-payments/pull/9558/commits/523df548779b4b9c99174dce8a0c19903468fcb7, I've reintroduced the `$_POST` checks that were removed in https://github.com/Automattic/woocommerce-payments/pull/9464 for the phone number field along with other metadata fields. I tested it by setting a breakpoint [here](https://github.com/Automattic/woocommerce-payments-server/blob/f041e28b9dc1aae1d764d89cf9704565791a680b/server/wp-content/rest-api-plugins/endpoints/wcpay/woopay/service/class-woopay-service.php#L290) and making sure following metadata fields are not null.

```
woopay_phone
source_url
is_blocks
viewport
```

a-danae committed:
Merge branch 'develop' into fix/woopay-user-creation-in-shortcode-checkout

a-danae reviewed (approved):
Things look good to me 👍 Thanks for iterating on this @malithsen !

> I tested it by setting a breakpoint [here](https://github.com/Automattic/woocommerce-payments-server/blob/f041e28b9dc1aae1d764d89cf9704565791a680b/server/wp-content/rest-api-plugins/endpoints/wcpay/woopay/service/class-woopay-service.php#L290) and making sure following metadata fields are not null.

And great catch identifying those other empty properties!

What I tested:
✅ The WooPay user gets created when using the "Save my info" checkbox in the shortcode checkout.
✅  The WooPay user gets created when using the "Save my info" checkbox in the blocks checkout.
✅ There are no regressions for #9235. The phone country code persists under the "Save my info" phone field on blur.
✅ There are no regressions for #9431. The phone field under "Save my info" component is pre-filled with the one used in the billing form.
✅ The code changes look good to me.

Thanks again! Happy to see this merged.
